Intrusive list safety

joe-redpanda opened the pull request:
This PR deletes the copy and move semantics of intrusive_list_hook and safe_intrusive_list_hook by default to prevent subtle bugs which arise from using copy and move semantics on intrusive list elements concurrently with iterators on their derived intrusive lists.

Intrusive lists have subtle behavior when copied or moved.

As of writing, list_base_hook has copy semantics where copy construction default initializes the new list hooks, and copy assignment is a noop.

list_base_hook has deleted move semantics.

Given these and an object which contains an intrusive_list_hook, the containing object's default move constructor will move its contained data as normal, but invoke the above copy semantics for intrusive list hook.

A user might expect that moving an intrusive list element would also move its linkages to the moved-into-object, but this does not occur.

This can be a problem in any container which does not offer pointer stability (including chunked_vector and chunked_hash_map). If one of these collections owns the memory of intrusive list elements, then updates to these containers while concurrently iterating over a derived intrusive list can lead to dangling pointers, iterator invalidation, or reordering of intrusive list elements.

This pr answers this by deleting copy and move semantics by default such that intrusive list hooks may not be used naively with containers for which this could be a problem.

Note: this does not answer iterator invalidation and its corresponding dangling reference issues, which occurs when an intrusive list element is destructed while an intrusive list iterator currently points to that element.

See https://godbolt.org/z/7hGxsTzYv for a demonstration of reordering and iterator invalidation.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [x] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Improvements

* Increases safety on our intrusive list wrappers

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda commented:
Should this be backported?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a8bca560-01be-49d5-9a4c-ec8075b933a7-->
<!--TEST RESULTS-->
<!--4d2dbaf9-8fa7-4912-8f3f-de482307ac0b-->
<!--TEST RESULTS-->
<!--dab3de27-1b3f-4d18-bbe2-25638ffdc6a2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72557">build#72557</a></summary>


| test_class               | test_method                         | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                            |
|:-------------------------|:------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePostRestartProbeTest | post_restart_probe_test             | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72557#01995e97-f9da-4962-b04a-42835aaf6acf | FLAKY         | 19/21    | upstream reliability is '99.57627118644068'. current run reliability is '90.47619047619048'. drift is 9.10008 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodePostRestartProbeTest&test_method=post_restart_probe_test          |
| PartitionBalancerTest    | test_recovery_mode_rebalance_finish | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72557#01995e97-f9de-4dd6-9c65-39ec81b1d1d4 | FLAKY         | 16/21    | upstream reliability is '81.71521035598705'. current run reliability is '76.19047619047619'. drift is 5.52473 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_recovery_mode_rebalance_finish |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72557#01995e97-f9df-4b18-8ec8-d5ad76de8580 | FLAKY         | 20/21    | upstream reliability is '98.54014598540147'. current run reliability is '95.23809523809523'. drift is 3.30205 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72969">build#72969</a></summary>


| test_class                    | test_method                                         | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                           |
|:------------------------------|:----------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic                              | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/72969#019981f3-79dd-464a-bd5c-4b62c6fbbefd | FLAKY         | 15/21    | upstream reliability is '91.54929577464789'. current run reliability is '71.42857142857143'. drift is 20.12072 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                     |
| ShadowLinkingReplicationTests | test_replication_basic                              | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/72969#019981fe-719c-4d4a-baa9-96ba895b10fa | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                     |
| ClusterRateQuotaTest          | test_client_group_consume_rate_throttle_mechanism   | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/72969#019981f3-79dd-464a-bd5c-4b62c6fbbefd | FLAKY         | 14/21    | upstream reliability is '81.28119800332779'. current run reliability is '66.66666666666666'. drift is 14.61453 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism   |
| ClusterRateQuotaTest          | test_client_group_consume_rate_throttle_mechanism   | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/72969#019981fe-719c-4d4a-baa9-96ba895b10fa | FLAKY         | 13/21    | upstream reliability is '93.9423076923077'. current run reliability is '61.904761904761905'. drift is 32.03755 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism   |
| ClusterRateQuotaTest          | test_client_response_and_produce_throttle_mechanism | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/72969#019981f3-79e0-4a92-b56d-7e782f60156f | FLAKY         | 14/21    | upstream reliability is '79.30501930501931'. current run reliability is '66.66666666666666'. drift is 12.63835 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism |
| DisablingPartitionsTest       | test_disable                                        | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/72969#019981f3-79e6-4d07-9fd9-9358fe611a51 | FLAKY         | 16/21    | upstream reliability is '85.8085808580858'. current run reliability is '76.19047619047619'. drift is 9.6181 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73019">build#73019</a></summary>


| test_class                     | test_method                             | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:-------------------------------|:----------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterRateQuotaTest           | test_client_response_throttle_mechanism | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73019#01998387-6bba-45e9-a367-29d23020849c | FLAKY         | 15/21    | upstream reliability is '81.40929535232384'. current run reliability is '71.42857142857143'. drift is 9.98072 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism |
| DatalakeCustomPartitioningTest | test_many_partitions                    | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/73019#01998387-6bb5-49fb-916f-d22f1fcfbf08 | FLAKY         | 10/21    | upstream reliability is '100.0'. current run reliability is '47.61904761904761'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeCustomPartitioningTest&test_method=test_many_partitions          |
| DisablingPartitionsTest        | test_disable                            | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73019#01998392-c376-4152-8e75-5ad438cdc6b7 | FLAKY         | 15/21    | upstream reliability is '85.41176470588235'. current run reliability is '71.42857142857143'. drift is 13.98319 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                         |
</details>
<!--/body-->

dotnwat reviewed (commented):
why should we care if intrusive list order changes when the owning container moves its elements? in most cases (except for things like LRU) I'd expect that intrusive lists are typically treated as unordered collections. perhaps ordering is a property that leads to more predictable behavior, but I think it should be motivated.

did you consider a different approach where we built something like a `non_movable_hook` which wraps an intrusive hook but disables moves so that when it is embedded in a type, that type's moves are disabled?

joe-redpanda commented:
> why should we care if intrusive list order changes when the owning container moves its elements?

Predictability motivated.
The cost of preserving list order is low (verbose constructors + relink operations on some moves)
while the benefit is preventing potentially subtle bugs and investigations into those bugs.

> did you consider a different approach where we built something like a `non_movable_hook` which wraps an intrusive hook but disables moves so that when it is embedded in a type, that type's moves are disabled?

Yeah. I feel like its strange that intrusive list hooks have copy and move enabled, but I assumed that boost had a reason for this that I was missing.

Happy to swap this to a wrapper

WillemKauf commented:
This is cool and topical as i'm currently writing even more intrusive lists into the codebase.

Are the current uses unsafe or is this (deleting move semantics) a precautionary step? Is there any slack/JIRA context around this?

WillemKauf commented:
And if you want, feel free to add more safety here: https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_topics/level_one/compaction/meta.h

dotnwat commented:
> Happy to swap this to a wrapper

well, it's not so much a request as it is a conversation starter. this PR fixes existing cases, but it doesn't help prevent new cases from appearing. is the later something we care about / can achieve?

bashtanov commented:
+1 for fixing it at the hook level. Can we make two wrappers, one swapping on move and another non-movable, both not copyable?

> I feel like its strange that intrusive list hooks have copy and move enabled, but I assumed that boost had a reason for this that I was missing.

The [copy ctor comment](https://github.com/boostorg/intrusive/blame/23e7a9cbf54f69f2bdc22951af959af74b037d10/include/boost/intrusive/list_hook.hpp#L106-L109) is from 2007, move ctors were not in the language at the time.

joe-redpanda commented:
> This is cool and topical as i'm currently writing even more intrusive lists into the codebase.
> 
> Are the current uses unsafe or is this (deleting move semantics) a precautionary step? Is there any slack/JIRA context around this?

The context is messy, but largely it surrounds this pr: https://github.com/redpanda-data/redpanda/pull/27525

The long and short is "be careful with intrusive list iterators". Normally no sane person edits a list while iterating over it, but with intrusive lists you must mind both the list itself and also the end owner of the memory.

Intrusive list by default does not swap links when a list member is moved, which means containers which invoke move operations (chunked_vector, chunked_hash_map) may reorder a list under your nose

joe-redpanda commented:
> +1 for fixing it at the hook level. Can we make two wrappers, one swapping on move and another non-movable, both not copyable?
> 
> > I feel like its strange that intrusive list hooks have copy and move enabled, but I assumed that boost had a reason for this that I was missing.
> 
> The [copy ctor comment](https://github.com/boostorg/intrusive/blame/23e7a9cbf54f69f2bdc22951af959af74b037d10/include/boost/intrusive/list_hook.hpp#L106-L109) is from 2007, move ctors were not in the language at the time.

Nice sleuthing.
I'm in agreement on the approach. I'll drop these in utils.

@dotnwat short of a clang-tidy complaint on raw intrusive list hook use, wrappers are probably the best we can do. I suspect this + a quick PSA will be okay. I frequently fish in the code base for examples of feature usage, so I would guess that any dev that misses the PSA might sense that something is amiss when a search for "intrusive_list_hook" only turns up custom wrappers.

joe-redpanda commented:
```
//! <b>Effects</b>: Swapping two nodes swaps the position of the elements
   //!   related to those nodes in one or two containers. That is, if the node
   //!   this is part of the element e1, the node x is part of the element e2
   //!   and both elements are included in the containers s1 and s2, then after
   //!   the swap-operation e1 is in s2 at the position of e2 and e2 is in s1
   //!   at the position of e1. If one element is not in a container, then
   //!   after the swap-operation the other element is not in a container.
   //!   Iterators to e1 and e2 related to those nodes are invalidated.
```

I need to audit how we are handling the above. This implies that (sans locking) it is not safe to use an intrusive list iterator across a scheduling point.

This pr will end up mostly functionally the same because its overwhelmingly deleting move semantics.

joe-redpanda committed:
tracking_allocator::mem_tracker: remove move

This commit removes the move constructor from mem_tracker in
anticipation for removing the move semantics from intrusive_list_hook.

This move constructor was unused.

joe-redpanda committed:
cloud_storage::remote: remove unneeded hook reset

Unlinking an intrusive list node is sufficient to remove and destruct an
intrusive list element. This commit removes a superfluous reassignment
of the intrusive list hook in anticipation of removing copy and move
semantics from intrusive list hook.

joe-redpanda committed:
namespaced_cache_test: custom copy and move

This adds custom copy and move semantics to intrusive list elements
within namespaced_cache_tests in anticipation for removing copy and move
semantics from intrusive list hooks.

joe-redpanda committed:
intrusive_list_helpers: remove copy and move

Copying and moving elements which participate in intrusive lists
can subtly invalidate iterators to their intrusive lists. This commit
removes copy and move semantics by default to prevent iterator
invalidation.

Move semantics may be custom implemented with hook::swap_nodes, though
writes should take care to ensure their usage is safe.

joe-redpanda commented:
> And if you want, feel free to add more safety here: https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_topics/level_one/compaction/meta.h

Those are fine for now, the changes coming in this pr will automatically grant safety against move and copy problems, asynchronous iteration over an intrusive list those is not safe per removals though and I would discourage it

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73019-->
<!--d219569c-d27d-4014-8fca-0ee85ce179ba-->
<!--title-->
## Retry command for [Build#73019](https://buildkite.com/redpanda/redpanda/builds/73019)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/custom_partitioning_test.py::DatalakeCustomPartitioningTest.test_many_partitions@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
```
<!--/body-->

bashtanov commented on src/v/cluster/tests/namespaced_cache_test.cc:
```diff
@@ -24,6 +24,29 @@ struct fixture : public seastar_test {};
     explicit entry(ss::sstring k)
       : key(std::move(k)) {}
+
+    // typically elements with intrusive lists should not have copy
+    // constructors, fine here because this is simply used to populate a
+    // vector with n copies of an entry
```
As far as I understand we have exactly two usages where elements with hooks are stored in a container that does not guarantee pointer stability. And they are both in this file. I guess if there were any other usages likes they wouldn't compile.

What do you think about creating a special hook wrapper for these, movable and copyable? Move will swap hooks and copy will vassert if the hook is linked.

joe-redpanda commented:
Update with rebase, CI has been failing on flakey tests and some fixes have since gone through
